### PR TITLE
add dynamic errorhandling

### LINKED ISSUE #81: Allow static or dynamic error pages
The idea is to alter the filename of the request and remember, that it is an error url. and then reprocess it.
I am afraid this will require some refactoring 
I would suggest to use a new name scheme for error pages to allow displaying them in unauthorized mode and prevent calling them directly
like
error/404.* for a 404 error and the like.

What do you think?

### PR DESCRIPTION
**Pull request by HHHartmann**
You can define static files as errorhandlers based on errorcode and file extension.

The last commits add **support for six cars** (or more) without crash. I serve static files without coroutine now. Very memory effichient!

Tested all sample pages successfully.

Also fix and improve many small details:

### CODE REVIEW

**Comment by ATAMAH** [comment]
For starting: [List of MIME types](http://www.iana.org/assignments/media-types/media-types.xhtml)

You can't change "text/html" to just "html" and "application/json" to "json".

**Comment by HHHartmann** [comment]
Thanks for reviewing the changes.
httpserver-header.lc has to be called with the file extension as 3rd parameter and then changes it to the corrosponding MIME type. So "json" becomes "application/json" and "html" becomes "text/html".
Calling it with "application/json" would result in the fallback MIME type "text/plain".

File: http/cars.html [hunk 1]
```diff
@@ -1,21 +0,0 @@
-<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
```
Comment (marcoskirsch) [inline http/cars.html:1]: I'd like to keep this around, it's useful to show the limitations of the server. I'm all for teaching clients how to work around said limitations, but it should be in a separate file.

Someone was supposed to do a pull request in which the workaround is based on having the served page include javascript to request the car images serially. I can see us having both examples included.

Can you do a separate PR in which you show your cars.lua workaround and leave this one in place? I'd rather do one pull request per logical change, this one bundles several.

Thanks!

File: http/node_info.lua [hunk 2]
```diff
@@ -1,5 +1,5 @@
 local function sendAttr(connection, attr, val)
-   connection:send("<li><b>".. attr .. ":</b> " .. val .. "<br></li>\n")
+   connection:send("<li><b>".. attr .. ":</b> " .. (val or "nil") .. "<br></li>\n")
```
Comment (marcoskirsch) [inline http/node_info.lua:2]: this improvement could be its own PR as well.

File: httpserver-buffer.lua [hunk 3]
```diff
@@ -0,0 +1,48 @@
+-- httpserver-connection
```
Comment (marcoskirsch) [inline httpserver-buffer.lua:1]: this comment seems outdated...
Why rename the file, though? Can you explain the reasoning behind this?

**Review by marcoskirsch — CHANGES_REQUESTED** [review]

File: http/cars.html (hunk 1 above)
Comment (HHHartmann) [reply http/cars.html:1]: Hi,

glad you find some time to have a closer look.

The old cars.html example will work as well with the changes herein. But if it is to try the other approach I can put it back in.
I just created the cars.lua since I was too lazy to change the sources every time I want to test a different number of cars. It generates pretty much the same page as the cars.html.

As to the big PR:
I developed offline and then tried to separate the chenges inseparate checkins. They build up on each other. Cause of that it would be quite difficult to extract each feature/fix in a separate PR.
Id like to avoid that.

File: http/node_info.lua (hunk 2 above)
Comment (HHHartmann) [reply http/node_info.lua:2]: Will do a separate PR here.

File: httpserver-buffer.lua (hunk 3 above)
Comment (HHHartmann) [reply httpserver-buffer.lua:1]: ITs not a rename but an alternative implementation of the connection overlay.
I changed sending static files to not use a coroutine as that is quite memory consuming.
So to still be able to use the header.lua I needed a buffer to collect the separate sends in a string and then send them out.
I then only store the filename and the current position to send the chunks of the file.

**Comment by HHHartmann** [comment]
So I will close this PR and create separate ones
